python/lib/scheduler/dmod/scheduler/job/job.py

```diff
@@ -451,53 +521,28 @@ def allocation_service_names(self) -> Optional[Tuple[str]]:
         """
         pass
 
-    @property
     @abstractmethod
-    def allocations(self) -> Optional[Tuple[ResourceAllocation]]:
-        """
-        The resource allocations that have been allocated for this job.
-
-        Returns
-        -------
-        Optional[List[ResourceAllocation]]
-            The scheduler resource allocations for this job, or ``None`` if it is queued or otherwise not yet allocated.
-        """
+    def set_allocations(self, allocations: List[ResourceAllocation]):
         pass
 
-    @allocations.setter
     @abstractmethod
-    def allocations(self, allocations: List[ResourceAllocation]):
+    def set_data_requirements(self, data_requirements: List[DataRequirement]):
         pass
 
-    @property
     @abstractmethod
-    def cpu_count(self) -> int:
-        """
-        The number of CPUs for this job.
-
-        Returns
-        -------
-        int
-            The number of CPUs for this job.
-        """
+    def set_partition_config(self, part_config: PartitionConfig):
         pass
 
-    @property
     @abstractmethod
-    def data_requirements(self) -> List[DataRequirement]:
-        """
-        List of ::class:`DataRequirement` objects representing all data needed for the job.
+    def set_status(self, status: JobStatus):
+        pass
 
-        Returns
-        -------
-        List[DataRequirement]
-            List of ::class:`DataRequirement` objects representing all data needed for the job.
-        """
+    @abstractmethod
+    def set_status_phase(self, phase: JobExecPhase):
```

Comment:
The previous version of the code does a little more than just update the property. It also mutates other state, so of which is done though a subclass specific private helper function (`_reset_last_updated`).

https://github.com/NOAA-OWP/DMOD/blob/62f7fbdfcad308d55d12bb5139d157b9b6225917/python/lib/scheduler/dmod/scheduler/job/job.py#L1066-L1073

To be fair of the classes I refactored, I am the least happy with the `Job` and `JobImpl` classes. The `JobImpl` classes had several settable properties (i.e. `@<prop_name>.setter`) that mutated internal state. This concept doesnt really map will to `pydantic` when those properties are also fields. You will notice that I added `set_<prop_name>` method to `Job` and now raise deprecated warnings when setting the properties through `o.prop = some_value` (i.e. through `__setattr__`). Below is the [commit](https://github.com/NOAA-OWP/DMOD/commit/90eea5ba14f9d328290b691f8780f2a5856a32d5) message I wrote when introducing those changes. I think I better synthesized the problem I was trying to solve in that text:

> Pre-pydantic versions of Job class used property setters.
>
> ```python
>    @property
>    def prop(self) -> str: ...
>
>    @prop.setter
>    def prop(self, value: str): ...
> ```
>
> Some properties with property setters mutatated state beyond just the
associated property (e.g. reset `last_updated`). Pydantic will not allow
you to create `@prop.setter`'s. Also, from a usage perspective, it not
always clear if you can set a property without looking at documentation.
To get around this, setter methods have been added to the Job class
making it clear that you _can_ modify a properties value after
instantiation. This also allows updating other instance internal state
(i.e. `last_updated`). Likewise, to help us find and update portions of
code that still set Job properties using prop setters, a
DeprecationWarning is now thrown.
>
> To accomplish raising DeprecationWarnings, Job's `__setattr__` method
was overridden. In short, there is a dict mapping of field names to
associated setter methods that is called instead of `__setattr__`. As a
side effect of this, setter methods must set field values through the
instance `__dict__`. This avoids setter methods from raising a
DeprecationWarning.